### PR TITLE
Handle updates to table schema when using Storage API writes.

### PR DESCRIPTION
**Pull request by reuvenlax**

### CODE REVIEW

**Comment by kennknowles** [comment]
There's no description. Is this just a working draft?

**Comment by reuvenlax** [comment]
Draft

**Comment by reuvenlax** [comment]
R: @prodriguezdefino 
R: @yirutang

**Comment by yirutang** [comment]
R: @GaoleMeng

**Comment by yirutang** [comment]
R: @agrawal-siddharth FYI

**Comment by reuvenlax** [comment]
friendly ping!

**Comment by reuvenlax** [comment]
friendly ping

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 1]
```diff
@@ -488,6 +542,14 @@ long flush(
               return RetryType.RETRY_ALL_OPERATIONS;
             },
             c -> {
+              AppendRowsResponse response = Preconditions.checkStateNotNull(c.getResult());
+              if (autoUpdateSchema && response.hasUpdatedSchema()) {
```
Comment (prodriguezdefino) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:546]: do we get a similar information piece on schema updates when the insert fails? if so that info added to the failed insert rows could help on debugging.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 2]
```diff
@@ -385,34 +399,57 @@ public void process(
...
+                            // blocking.
+                            client -> runAsyncIgnoreFailure(closeWriterExecutor, client::close))
+                        .withAppendClient(datasetService, getOrCreateStream, false);
+                  }));
+      if (autoUpdateSchema && updatedSchema.read() != null) {
+        if (appendClientInfo.get().hasSchemaChanged(updatedSchema.read())) {
+          appendClientInfo.set(
+              AppendClientInfo.of(
+                  updatedSchema.read(), appendClientInfo.get().getCloseAppendClient()));
+          // TODO: invalidate?
```
Comment (prodriguezdefino) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:434]: it seems that the unsharded version of the writes do invalidate when schema has changed.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 3]
```diff
@@ -640,6 +678,18 @@ public void process(
         }
         appendSplitDistribution.update(numAppends);
 
+        if (updatedSchemaReturned.get() != null) {
+          // Update the table schema and clear the append client.
+          if (appendClientInfo.get().hasSchemaChanged(updatedSchemaReturned.get())) {
+            appendClientInfo.set(
+                AppendClientInfo.of(
+                    updatedSchemaReturned.get(), appendClientInfo.get().getCloseAppendClient()));
+            // TODO: invalidate?
```
Comment (prodriguezdefino) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:687]: same as in line 434

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsBeamRow.java [hunk 4]
```diff
@@ -61,9 +61,15 @@ public TableSchema getTableSchema() {
...
+    @SuppressWarnings("nullness")
     public StorageApiWritePayload toMessage(T element) {
       Message msg = BeamRowToStorageApiProto.messageFromBeamRow(descriptor, toRow.apply(element));
-      return new AutoValue_StorageApiWritePayload(msg.toByteArray());
+      return new AutoValue_StorageApiWritePayload(msg.toByteArray(), null);
+    }
+
+    @Override
+    public StorageApiWritePayload toMessage(TableRow tableRow, boolean respectRequired) {
+      throw new RuntimeException("Not supported");
```
Comment (prodriguezdefino) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsBeamRow.java:72]: Maybe I misunderstood the logic, but would it be possible to add a validation in BigQueryIO that schema update is not available for Beam Row payloads? 
If I'm not mistaken someone could configure the writes with `useBeamSchema()` and set also auto update of schemas and then their pipeline will fail with this runtime exception, when in fact we could have captured that in validation time.

**Review by prodriguezdefino — COMMENTED** [review]
left few comment nits, but overall looks good to me

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:546]: I don't believe we do.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:434]: I _think_ that put() will invalidate the old value

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:687]: Same - I believe that put() invalidates the old value

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsBeamRow.java (hunk 4 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsBeamRow.java:72]: done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 3 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java]: added invalidate

**Comment by prodriguezdefino** [comment]
thanks, lgtm

**Comment by prodriguezdefino** [comment]
LGTM

**Comment by reuvenlax** [comment]
Run Java_GCP_IO_Direct PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 5]
```diff
@@ -512,6 +581,18 @@ String retrieveErrorDetails(Iterable<AppendRowsContext> failedContext) {
...
+
+      void postFlush() {
+        // If we got a response indicating an updated schema, recreate the client.
+        if (updatedTableSchema != null
+            && this.appendClientInfo != null
+            && this.appendClientInfo.hasSchemaChanged(updatedTableSchema)) {
+          invalidateWriteStream();
+          appendClientInfo =
+              Preconditions.checkStateNotNull(getAppendClientInfo(false, updatedTableSchema));
+          updatedTableSchema = null;
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:593]: I am wondering if there will be races regarding this updatedSchema?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 6]
```diff
@@ -364,9 +395,37 @@ void invalidateWriteStream() {
...
+          StorageApiWritePayload payload,
+          OutputReceiver<BigQueryStorageApiInsertError> failedRowsReceiver)
+          throws Exception {
         maybeTickleCache();
         ByteString payloadBytes = ByteString.copyFrom(payload.getPayload());
+        if (autoUpdateSchema) {
+          if (appendClientInfo == null) {
+            appendClientInfo = getAppendClientInfo(true, null);
+          }
+          @Nullable TableRow unknownFields = payload.getUnknownFields();
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:408]: unknownFields are values that were unknown to the prior conversion step. These fields may be known to the writing step (since it gets schema updates back from Vortex) so we can't ignore them here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AppendClientInfo.java [hunk 7]
```diff
@@ -28,35 +35,111 @@
...
+    if (client != null) {
+      getCloseAppendClient().accept(client);
+    }
+  }
+
+  boolean hasSchemaChanged(TableSchema updatedTableSchema) {
+    return updatedTableSchema.hashCode() != getTableSchema().hashCode();
+  }
+
+  public ByteString encodeUnknownFields(TableRow unknown, boolean ignoreUnknownValues)
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AppendClientInfo.java:115]: No - these are fields that are unknown to the prior step. They make actually end up being known to the current step due to the updated schema.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 8]
```diff
@@ -385,43 +399,68 @@ public void process(
...
+                                          // Remove the pin that is "owned" by the cache.
+                                          client.unpin();
+                                          client.close();
+                                        }))
+                            .withAppendClient(datasetService, getOrCreateStream, false);
+                    // This pin is "owned" by the cache.
+                    Preconditions.checkStateNotNull(info.getStreamAppendClient()).pin();
+                    return info;
+                  }));
+      TableSchema updatedSchemaValue = updatedSchema.read();
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:439]: updatedSchema can be null?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 5 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:593]: what race do you envision?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 8 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:439]: updatedSchema is a state variable. If it has never been set, then it will return null.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java [hunk 9]
```diff
@@ -479,15 +502,41 @@ public static DynamicMessage messageFromTableRow(
         }
       }
 
+      if (unknownFields != null) {
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:505]: Is this file the conversion stage? 

If user has input ABC and the original schema is AB but later on the schema is updated to ABC, will L496 fail the check?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java [hunk 10]
```diff
@@ -1035,6 +1038,78 @@ public void testRejectUnknownNestedFieldF() throws Exception {
...
+    TableRowToStorageApiProto.messageFromTableRow(
+        schemaInformation, descriptor, row, true, false, ignored);
+    assertEquals(BASE_TABLE_ROW.getF().size() + 1, ignored.getF().size());
+    assertEquals("foobar", ignored.getF().get(BASE_TABLE_ROW.getF().size()).getV());
+  }
+
+  @Test
+  public void testIgnoreUnknownNestedField() throws Exception {
+    TableRow rowNoF = new TableRow();
+    rowNoF.putAll(BASE_TABLE_ROW_NO_F);
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java:1086]: Just try to understand the test, so Beam accepts two format of TableRow? One is "F"->List of values and the other is List(field_name, field_value), and they can be mixed up as nested field value to the same TableRow struct?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 5 above)
Comment (yirutang) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:593]: Some responses are coming back and the updatedTableSchema is being updated L568, which in race with the postFlush here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java (hunk 9 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProto.java:505]: no, because the caller (StorageApiDynamicDestinationsTableRow.java:151) passes in true if autoSchemaUpdates==true

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java (hunk 10 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java:1086]: correct. This is the unfortunate history of TableRow (which long predated Beam)

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java [hunk 11]
```diff
@@ -1035,6 +1038,78 @@ public void testRejectUnknownNestedFieldF() throws Exception {
...
+    TableRow unknown = new TableRow();
+    TableRowToStorageApiProto.messageFromTableRow(
+        schemaInformation, descriptor, topRow, true, false, unknown);
+    assertEquals(3, unknown.size());
+    assertEquals("foobar", unknown.get("unknowntop"));
+    assertEquals(1, ((TableRow) unknown.get("nestedvalue1")).size());
+    assertEquals(1, ((TableRow) unknown.get("nestedvaluenof1")).size());
+    assertEquals(
+        "foobar",
+        ((TableRow) unknown.get("nestedvalue1")).getF().get(BASE_TABLE_ROW.getF().size()).getV());
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java:1112]: In this case, the unknown value is at the last offset, is there elements ahead of it? What if the unknown field is in the middle of the array list?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 5 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:593]: postFlush is called only after all futures have completed (in flushAll), so we would not expect any more callbacks. Also note that RetryManager calls these response callbacks in the primary thread (RetryManager.await() calls the callbacks), so the callbacks here are not being called asynchronously.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java [hunk 12]
```diff
@@ -1851,6 +1859,197 @@ public TableRow apply(Long input) {
...
+                    new UpdateTableSchemaDoFn(
+                        Duration.standardSeconds(5), tableSchemaUpdated, fakeDatasetService)))
+            .setCoder(TableRowJsonCoder.of());
+
+    tableRows.apply(
+        BigQueryIO.writeTableRows()
+            .to(tableRef)
+            .withMethod(method)
+            .withCreateDisposition(BigQueryIO.Write.CreateDisposition.CREATE_NEVER)
+            .ignoreUnknownValues()
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:1996]: nit: Add a test for ignoreUnknownValue(false) and withAutoSchemaUpdate(true)

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java [hunk 13]
```diff
@@ -1851,6 +1859,197 @@ public TableRow apply(Long input) {
...
+
+    p.run();
+
+    Iterable<TableRow> expectedDroppedValues =
+        LongStream.range(0, 6)
+            .mapToObj(getRowSet)
+            .map(tr -> filterUnknownValues(tr, tableSchema.getFields()))
+            .collect(Collectors.toList());
+    Iterable<TableRow> expectedFullValues =
+        LongStream.range(6, 10).mapToObj(getRowSet).collect(Collectors.toList());
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:2009]: Should this be 6,9 and above 0,5?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java [hunk 14]
```diff
@@ -17,23 +17,50 @@
...
+import java.util.function.Function;
+import javax.annotation.Nullable;
 
 /**
  * Takes in an iterable and batches the results into multiple ProtoRows objects. The splitSize
  * parameter controls how many rows are batched into a single ProtoRows object before we move on to
  * the next one.
  */
 class SplittingIterable implements Iterable<ProtoRows> {
+  interface ConvertUnknownFields {
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java:35]: I somehow couldn't find the implementation of this?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java [hunk 15]
```diff
@@ -57,7 +84,37 @@ public ProtoRows next() {
...
+          if (autoUpdateSchema) {
+            try {
+              @Nullable TableRow unknownFields = payload.getUnknownFields();
+              if (unknownFields != null) {
+                // Protocol buffer serialization format supports concatenation. We serialize any new
+                // "known" fields
+                // into a proto and concatenate to the existing proto.
+                try {
+                  byteString =
+                      byteString.concat(
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java:96]: Maybe I missed something, so every time we will concat the unknown fields with the existing byte string? Then what's the difference between this and passing down the entire message? Maybe the unknownFieldsToMessage will filter something out? But I don't see it has a schema.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinations.java [hunk 16]
```diff
@@ -35,6 +35,8 @@
 
     StorageApiWritePayload toMessage(T element) throws Exception;
 
+    StorageApiWritePayload toMessage(TableRow tableRow, boolean respectRequired) throws Exception;
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinations.java:38]: How is this relevant to this change?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 17]
```diff
@@ -512,6 +581,18 @@ String retrieveErrorDetails(Iterable<AppendRowsContext> failedContext) {
             .map(StackTraceElement::toString)
             .collect(Collectors.joining("\n"));
       }
+
+      void postFlush() {
+        // If we got a response indicating an updated schema, recreate the client.
+        if (updatedTableSchema != null
+            && this.appendClientInfo != null
+            && this.appendClientInfo.hasSchemaChanged(updatedTableSchema)) {
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:589]: Not sure if this is expensive to perform, theoretically it is not needed since when setting updatedTableSchema it is already checked. 

We also provided something here:
https://github.com/googleapis/java-bigquerystorage/blob/main/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/StreamWriter.java#L157 

If that field is not null, it means a new schema appeared. We do the comparison based on timestamp.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 18]
```diff
@@ -652,6 +692,18 @@ public void process(
         }
         appendSplitDistribution.update(numAppends);
 
+        if (updatedSchemaReturned.get() != null) {
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:695]: Wondering if you can directly use the one on StreamWriter:
https://github.com/googleapis/java-bigquerystorage/blob/main/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/StreamWriter.java#L157

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java (hunk 14 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java:35]: This is a Java functional interface - any matching lambda will conform. e.g. you can pass in (tableRow, ignore) -> {} and this will conform to the interface

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java (hunk 15 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java:96]: The prior convert message only includes fields known to it in the proto it generates. It can't include fields it doesn't know about as they would have to be in the proto descriptor (and it can't use the proto's unknown field set as that requires field ids, which is not known yet).

Therefore the incoming byteString contains only fields that were known to the convert stage, and all other fields are put into the json unknownFields object. What we are doing here is taking advantage of the fact that the write step has a more up-to-date view on the schema, so we walk over the unknownFields json and extract whatever fields are now known (which might still be only a subset of the remaining fields). We then convert those unknownFields to a proto, and concatenate the two protos.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinations.java (hunk 16 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinations.java:38]: We can't simply take the unknownFields and convert them to a proto, as there may be missing required fields (because those fields are in the original proto). We need a way to do the conversion without enforcing nullability.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 18 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:695]: Maybe - is there an advantage to using that over the one in the response?

Comment (yirutang) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:695]: There is a contract that it will only appear when a schema update happened (for once), which is our JsonWriter depending on, so you don't need to do the comparisons at all.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java (hunk 15 above)
Comment (yirutang) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java:96]: Not an expert in modern Java... Where is the implementation of unknownFieldsToMessage.convert? How could only convert the "known" unknown fields?

Comment (yirutang) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SplittingIterable.java:96]: Yes, code looks good here.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java (hunk 11 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/TableRowToStorageApiProtoTest.java:1112]: This is an unordered map, so there should be no offsets.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java (hunk 12 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:1996]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 18 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:695]: Done

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java (hunk 13 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java:2026]: (6,10) is open on the upper bound (i.e. it's [6, 10) )

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 17 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java]: removed this
